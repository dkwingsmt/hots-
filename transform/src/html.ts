import _ from 'lodash';
import * as parse5 from 'parse5';

import { IPageInfo, Node } from './const';
import { generateBbsCode } from './generate';
import { translateTree } from './translate';
import { isElement, isParentNode, isTextNode } from './utils';

function getArticleTreeTraverse(tree: parse5.DefaultTreeNode) {
  if (tree.nodeName === 'article') {
    return tree;
  }
  if (tree.nodeName === 'blockquote') {
    return tree;
  }
  if (isElement(tree)) {
    const classAttr = tree.attrs.find((v: parse5.Attribute) => v.name === 'class');
    if (classAttr && classAttr.value.includes('notes-detail')) {
      return tree;
    }
    if (classAttr && classAttr.value.includes('news_area')) {
      return tree;
    }
    if (classAttr && classAttr.value.includes('article-container')) {
      return tree;
    }
    const idAttr = tree.attrs.find((v: parse5.Attribute) => v.name === 'id');
    if (idAttr && idAttr.value === 'mainNews') {
      return tree;
    }
  }
  if (!isParentNode(tree)) {
    return undefined;
  }
  let result;
  _.forEach(tree.childNodes, (node: parse5.DefaultTreeNode) => {
    if (node) {
      const found = getArticleTreeTraverse(node);
      if (found) {
        result = found;

        return false; // break
      }
    }
  });

  return result;
}

function getArticleTree(htmlText: string) {
  const htmlTree = <parse5.DefaultTreeDocument>parse5.parse(htmlText);
  const articleTree = getArticleTreeTraverse(htmlTree);
  if (!articleTree) {
    throw new Error('Can\'t find the article.');
  }

  return articleTree;
}

export function standardizeTree(node: parse5.DefaultTreeNode): Node {
  if (isTextNode(node)) {
    return {
      kind: 'text',
      text: node.value,
    };
  }

  return {
    kind: 'parent',
    tag: node.nodeName,
    children: isParentNode(node) ? node.childNodes.map(standardizeTree) : [],
    attrs: isElement(node) ? node.attrs : [],
  };
}

function transformNgaNode(tree: Node): string {
  return generateBbsCode(tree);
}

export function pageToBbsCode({ htmlText, url }: IPageInfo, { doTranslate }: {doTranslate: boolean}) {
  const tree = standardizeTree(getArticleTree(htmlText));
  const translatedTree = doTranslate ? translateTree(tree) : tree;
  const sourceStr = url ? `[quote]英文日志：${url}
[/quote]
` : '';

  return `[quote]转载请注明本帖来源NGA[s:a2:poi]
[/quote]
${sourceStr}${transformNgaNode(translatedTree)}`;
}
