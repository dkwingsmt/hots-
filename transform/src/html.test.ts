import * as parse5 from 'parse5';

import { standardizeTree } from './html';

describe('test standardizeTree()', () => {
  test('result must match', () => {
    const node = <parse5.DefaultTreeDocumentFragment>parse5.parseFragment(
      `<b style="1" disabled=1> <i /> 123 <em><u>hey</u></em></b>`,
    );
    expect(standardizeTree(node.childNodes[0])).toMatchSnapshot();
  });
});
