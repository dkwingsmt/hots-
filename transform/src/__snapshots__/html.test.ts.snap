// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test standardizeTree() result must match 1`] = `
Object {
  "attrs": Array [
    Object {
      "name": "style",
      "value": "1",
    },
    Object {
      "name": "disabled",
      "value": "1",
    },
  ],
  "children": Array [
    Object {
      "kind": "text",
      "text": " ",
    },
    Object {
      "attrs": Array [],
      "children": Array [
        Object {
          "kind": "text",
          "text": " 123 ",
        },
        Object {
          "attrs": Array [],
          "children": Array [
            Object {
              "attrs": Array [],
              "children": Array [
                Object {
                  "kind": "text",
                  "text": "hey",
                },
              ],
              "kind": "parent",
              "tag": "u",
            },
          ],
          "kind": "parent",
          "tag": "em",
        },
      ],
      "kind": "parent",
      "tag": "i",
    },
  ],
  "kind": "parent",
  "tag": "b",
}
`;
